client/findproviders.go

```diff
@@ -56,23 +56,30 @@ func (fp *Client) FindProvidersAsync(ctx context.Context, key cid.Cid) (<-chan F
 	if err != nil {
 		return nil, err
 	}
+
 	parsedRespCh := make(chan FindProvidersAsyncResult, 1)
 	go func() {
 		defer close(parsedRespCh)
-		if ctx.Err() != nil {
-			return
-		}
-		protoAsyncResp, ok := <-protoRespCh
-		if !ok {
-			return
-		}
+
+		// TODO wrap in a for loop after fixing https://github.com/ipld/go-ipld-prime/issues/374
 		var parsedAsyncResp FindProvidersAsyncResult
-		parsedAsyncResp.Err = protoAsyncResp.Err
-		if protoAsyncResp.Resp != nil {
-			parsedAsyncResp.AddrInfo = parseFindProvidersResponse(protoAsyncResp.Resp)
+		select {
+		case <-ctx.Done():
+			parsedAsyncResp.Err = ctx.Err()
```

Comment:
In the event of context cancellation, we should just return without writing more to the channel. So just put a `return` statement here. 

Recall, the goal is to make it safe for the user to cancel the context and stop reading from the channel. While this current implementation won't block, because you have a buffer in the channel, this won't work once you add the for loop.